subprojects/publish/src/main/java/org/gradle/api/publish/internal/metadata/ModuleMetadataSpecBuilder.java

```diff
@@ -385,47 +356,28 @@ private List<ModuleMetadataSpec.DependencyConstraint> dependencyConstraintsFor(S
             return emptyList();
         }
-        VariantVersionMappingStrategyInternal versionMappingStrategy = versionMappingStrategyFor(variant);
+        VariantDependencyResolver dependencyResolver = variantDependencyResolverFor(variant);
         ArrayList<ModuleMetadataSpec.DependencyConstraint> dependencyConstraints = new ArrayList<>();
         for (DependencyConstraint dependencyConstraint : variant.getDependencyConstraints()) {
             dependencyConstraints.add(
-                dependencyConstraintFor(dependencyConstraint, versionMappingStrategy, variant.getName())
+                dependencyConstraintFor(dependencyConstraint, dependencyResolver, variant.getName())
             );
         }
         return dependencyConstraints;
     }
 
-    private ModuleMetadataSpec.DependencyConstraint dependencyConstraintFor(DependencyConstraint dependencyConstraint,
-                                                                            VariantVersionMappingStrategyInternal variantVersionMappingStrategy,
-                                                                            String variant) {
-        String group;
-        String module;
-        String resolvedVersion = null;
-        Path identityPath = null;
-        if (dependencyConstraint instanceof DefaultProjectDependencyConstraint) {
-            DefaultProjectDependencyConstraint dependency = (DefaultProjectDependencyConstraint) dependencyConstraint;
-            ProjectDependency projectDependency = dependency.getProjectDependency();
-            identityPath = ((ProjectDependencyInternal) projectDependency).getIdentityPath();
-            ModuleVersionIdentifier identifier = projectDependencyResolver.resolve(ModuleVersionIdentifier.class, identityPath);
-            group = identifier.getGroup();
-            module = identifier.getName();
-            resolvedVersion = identifier.getVersion();
-        } else {
-            group = dependencyConstraint.getGroup();
-            module = dependencyConstraint.getName();
-        }
-        ModuleVersionIdentifier resolvedVersionId = variantVersionMappingStrategy != null
-            ? variantVersionMappingStrategy.maybeResolveVersion(group, module, identityPath)
-            : null;
-        String effectiveGroup = resolvedVersionId != null ? resolvedVersionId.getGroup() : group;
-        String effectiveModule = resolvedVersionId != null ? resolvedVersionId.getName() : module;
-        String effectiveVersion = resolvedVersionId != null ? resolvedVersionId.getVersion() : resolvedVersion;
+    private ModuleMetadataSpec.DependencyConstraint dependencyConstraintFor(
+        DependencyConstraint dependencyConstraint,
+        VariantDependencyResolver dependencyResolver,
+        String variant
+    ) {
+        VariantDependencyResolver.Coordinates identifier = dependencyResolver.resolveComponentCoordinates(dependencyConstraint);
         return new ModuleMetadataSpec.DependencyConstraint(
```

Comment:
💭 Similar to `DependencyCoordinates` above.

Reply:
Discussed above